core/app/models/spree/order.rb

```diff
@@ -648,11 +648,11 @@ def cart_promo_total
         sum(:amount)
     end
 
-    def has_free_shipping?
-      promotions.
-        joins(:promotion_actions).
-        where(spree_promotion_actions: { type: 'Spree::Promotion::Actions::FreeShipping' }).
-        exists?
+    def free_shipping?
+      shipment_adjustments.
+        joins('LEFT OUTER JOIN spree_promotion_actions ON spree_promotion_actions.id = spree_adjustments.source_id').
```

Comment:
Note: can't use rails `.join` method because it's a polymorphic association.

Reply:
@KacperMekarski please do not change the method name, it will breaking change which we would not want to introduce

Reply:
@KacperMekarski also 

1. please do not use raw SQL (this will raise a warning in future rails versions)
2. For DB table names please use `Spree::PromotionAction.table_name` do not hardcode it

Reply:
You can actually use `left_outer_joins` https://blog.bigbinary.com/2016/03/24/support-for-left-outer-joins-in-rails-5.html

Reply:
@damianlegawiec 
I switched method name back to `has_free_shipping?`

However relating to raw SQL, I can't use `.left_outer_join` on polymorphic association just like that because I have to define where `source_id` as foreign key should lead (it's polymorphic so it leads to credit_card, payment, refund and finally **promotion_action**). We have two options:
- define association to promotion_action by hand in adjustment model (which I did now)
- use raw SQL (old solution)
Sources:
http://nerds.intuo.io/2018/03/01/eagerload-polymorphic-associations-in-rails.html
https://whatraghulearned.wordpress.com/tag/rails-polymorphic-association-left-outer-join/
https://blog.bigbinary.com/2020/01/07/rails-multiple-polymorphic-joins.html